base/multidimensional.jl

```diff
@@ -14,19 +14,34 @@ export CartesianIndex, CartesianRange
 linearindexing{A<:BitArray}(::Type{A}) = LinearFast()
 
 # CartesianIndex
-immutable CartesianIndex{N}
+immutable CartesianIndex{N} <: AbstractCartesianIndex{N}
     I::NTuple{N,Int}
     CartesianIndex(index::NTuple{N,Integer}) = new(index)
 end
 
 CartesianIndex{N}(index::NTuple{N,Integer}) = CartesianIndex{N}(index)
-@generated function (::Type{CartesianIndex{N}}){N}(index::Integer...)
+(::Type{CartesianIndex})(index::Integer...) = CartesianIndex(index)
+(::Type{CartesianIndex{N}}){N}(index::Integer...) = CartesianIndex(index)
+# Allow passing tuples smaller than N
+@generated function (::Type{CartesianIndex{N}}){N,M}(index::NTuple{M,Integer})
```

Comment:
Turns out we can get rid of the `@generated` function here too, see 1ed27d621523aad118f9d8459460551881ac9beb (added to #15030). Then this call would be
```jl
(::Type{CartesianIndex{N}}){N,M}(index::NTuple{M,Integer}) = CartesianIndex{N}(fill_to_length(index, 1, Val{N}))
```
This is also a more composable approach, since `fill_to_length` may be useful in a variety of contexts.

For reference:
```jl
julia> @code_llvm fill_to_length((1,2,3), -1, Val{5})

define void @julia_fill_to_length_23994([5 x i64]* sret, [3 x i64]*, i64, %jl_value_t*) #0 {
top:
  %4 = call %jl_value_t*** @jl_get_ptls_states()
  %5 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 0
  %6 = load i64, i64* %5, align 8
  %7 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 1
  %8 = load i64, i64* %7, align 8
  %9 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 2
  %10 = load i64, i64* %9, align 8
  %11 = insertvalue [5 x i64] undef, i64 %6, 0
  %12 = insertvalue [5 x i64] %11, i64 %8, 1
  %13 = insertvalue [5 x i64] %12, i64 %10, 2
  %14 = insertvalue [5 x i64] %13, i64 %2, 3
  %15 = insertvalue [5 x i64] %14, i64 %2, 4
  store [5 x i64] %15, [5 x i64]* %0, align 8
  ret void
}
```
compare against
```jl
julia> @generated function fill_to_length_gen{M,N}(t::NTuple{M}, val, ::Type{Val{N}})
           M > N && error("input tuple has length $M, asked for $N")
           args = [d <= M ? :(t[$d]) : :(val) for d = 1:N]
           :(tuple($(args...)))
       end

julia> @code_llvm fill_to_length_gen((1,2,3), -1, Val{5})

define void @julia_fill_to_length_gen_23846([5 x i64]* sret, [3 x i64]*, i64, %jl_value_t*) #0 {
top:
  %4 = call %jl_value_t*** @jl_get_ptls_states()
  %5 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 0
  %6 = load i64, i64* %5, align 8
  %7 = insertvalue [5 x i64] undef, i64 %6, 0
  %8 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 1
  %9 = load i64, i64* %8, align 8
  %10 = insertvalue [5 x i64] %7, i64 %9, 1
  %11 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 2
  %12 = load i64, i64* %11, align 8
  %13 = insertvalue [5 x i64] %10, i64 %12, 2
  %14 = insertvalue [5 x i64] %13, i64 %2, 3
  %15 = insertvalue [5 x i64] %14, i64 %2, 4
  store [5 x i64] %15, [5 x i64]* %0, align 8
  ret void
}